modules/index/src/main/java/org/apache/ignite/internal/idx/InternalSortedIndexImpl.java

```diff
@@ -0,0 +1,322 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.idx;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.BitSet;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.apache.ignite.internal.schema.BinaryRow;
+import org.apache.ignite.internal.schema.Column;
+import org.apache.ignite.internal.storage.index.IndexRow;
+import org.apache.ignite.internal.storage.index.IndexRowPrefix;
+import org.apache.ignite.internal.storage.index.SortedIndexColumnDescriptor;
+import org.apache.ignite.internal.storage.index.SortedIndexDescriptor;
+import org.apache.ignite.internal.storage.index.SortedIndexStorage;
+import org.apache.ignite.internal.table.StorageRowListener;
+import org.apache.ignite.internal.table.TableImpl;
+import org.apache.ignite.internal.table.TableRow;
+import org.apache.ignite.internal.util.Cursor;
+import org.apache.ignite.lang.IgniteInternalException;
+import org.apache.ignite.lang.IgniteUuid;
+import org.apache.ignite.table.Tuple;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Internal index manager facade provides low-level methods for indexes operations.
+ */
+public class InternalSortedIndexImpl implements InternalSortedIndex, StorageRowListener {
+    private final String name;
+
+    private final TableImpl tbl;
+
+    private final SortedIndexStorage store;
+
+    private final SortedIndexDescriptor desc;
+
+    /**
+     * Create sorted index.
+     */
+    public InternalSortedIndexImpl(String name, SortedIndexStorage store, TableImpl tbl) {
+        this.name = name;
+        this.store = store;
+        this.tbl = tbl;
+
+        desc = store.indexDescriptor();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public String name() {
+        return name;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public String tableName() {
+        return tbl.name();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public IgniteUuid tableId() {
+        return tbl.tableId();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public List<Column> columns() {
+        return store.indexDescriptor().columns().stream()
+                .map(SortedIndexColumnDescriptor::column)
+                .collect(Collectors.toList());
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public Cursor<Tuple> scan(Tuple low, Tuple up, byte scanBoundMask, BitSet proj) {
```

Comment:
scanBoundMask - not used?

Reply:
As I see it, the SQL engine is not ready to use this optimization.